### ISSUE TITLE
Stress Testing GP Expressiveness

### ISSUE THREAD

**Issue by Duane321** [cornellius-gp/gpytorch#559]
Hi,

I'm working on a problem that I think presses GPs on their expressiveness. To demo it, I've generated a synthetic dataset that mimics the mechanism I'm trying to model. Ultimately, I'm hoping to gather feedback as to how to best use GPyTorch for a problem like this. 

With that, I'll load in some data. It's attached as:

[Expressive_GP_data.csv.zip](https://github.com/cornellius-gp/gpytorch/files/2949006/Expressive_GP_data.csv.zip)

But first, some helper functions:

```
def clean_ax(ax):
    """Used to clean up the axis of a matplotlib axes"""
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)
    ax.grid()
    
def generate_circle_features(x, period):
    """Fitting an RBF on these features will learn a periodic function."""
    return np.stack([np.sin(x*2*np.pi/period),
                     np.cos(x*2*np.pi/period)], axis=1)

def make_cubic_spline_basis(x, knots):
    """Forms a cubic spline basis. In other words, regression on what is returned here will fit a cubic spline."""
    X_list = []

    for e in range(4):
        X_list.append(np.power(x,e))

    for k in knots:
        X_list.append([np.power(max(xi-k,0),3) for xi in x])
        
    return np.stack(X_list, axis=1)
```

Now, let's read/inspect the data:

```
data = pd.read_csv('Data/Expressive_GP_data.csv')
data.head()
```
![image](https://user-images.githubusercontent.com/19956442/54079663-97737f80-4295-11e9-9350-3f0eec965e60.png)

Which looks like:

```
fig, axarr = plt.subplots(6, 1, sharex=True, figsize=(12,8))

for i in range(1, data.shape[1]):
    ax = axarr[i-1]
    ax.plot(data['x'], data.iloc[:,i])
    ax.set_title(data.columns[i])
    clean_ax(ax)
    
fig.tight_layout()
```
![image](https://user-images.githubusercontent.com/19956442/54079669-b1ad5d80-4295-11e9-95b4-e8adca00625e.png)

With that, I can tell you y was constructed as:

    y = spline_m*(cyclic_third + cyclic_10th + spline_t + noise)
    
Further,

1. x is 2K points evenly spaced between 0 and 1
2. spline_m is a cubic spline with knots at 1/3 and 2/3.
3. cyclic_third is a sampled RBF function on unit circle features with period 1/3 (so it's an RBF on [np.sin(x \* pi \* 2 \* 3), np.cos(x \* pi \* 2 \*10)]
4. cyclic_10th is the same thing, but with period 1/10
5. spline_t is a spline with knots at 1/3 and 2/3.
6. noise is samples of a random normal with mean zero and standard dev of 0.02. Notice it's multiplied by spline_m

So, this is quite a weird structure and something that a GP would be really well suited to tackling.

To do that, we'll need to construct X, which will have:

1. the cubic spline basis with knots at 1/3 and 2/3. So this is a column of 1's, x, x^2, x^3, max((x-1/3)^3,0), max((x-2/3)^3,0)
2. sin and cos from a unit circle with period 1/3. An additional 2 features for a period of 1/10.

Since the data we generated from these features, this should be everything we need. Let's set up X and y:

```
x = data['x'].values

X = [make_cubic_spline_basis(x, [1/3,2/3]),
     generate_circle_features(x, 1/3),
     generate_circle_features(x, 1/10)]
X = np.concatenate(X,axis=1)

# X = [make_cubic_spline_basis(x, [1/3,2/3])[:,:2],
#      generate_circle_features(x, 1/3),
#      generate_circle_features(x, 1/10)]
# X = np.concatenate(X,axis=1)

X = torch.from_numpy(X).to(torch.float64)
y = torch.from_numpy(data['y'].values).to(torch.float64)
```

Next, we fit a model:

```
class GP(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(GP, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        
        spline_indices = [0, 1, 2, 3, 4, 5]
        self.ker_linear_t = gpytorch.kernels.LinearKernel(num_dimensions=len(spline_indices), 
                                                          active_dims=spline_indices)
        self.ker_linear_m = gpytorch.kernels.LinearKernel(num_dimensions=len(spline_indices), 
                                                          active_dims=spline_indices)
        
        self.ker_cyc_half = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(active_dims=[6,7]))
        self.ker_cyc_10th = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(active_dims=[8,9]))
        
        noise_vars = np.array([0.02**2]*len(x))
        noise_vars = torch.from_numpy(noise_vars).to(torch.float64)
        self.ker_noise = gpytorch.kernels.WhiteNoiseKernel(noise_vars)
        
        self.covar_module = self.ker_linear_m*(self.ker_linear_t + self.ker_cyc_half + self.ker_cyc_10th + self.ker_noise)
    
    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
    
likelihood = gpytorch.likelihoods.GaussianLikelihood()
likelihood.initialize(**{'noise_covar.raw_noise':0.00}) 

model = GP(X, y, likelihood)

learning_rate = .1
model.train()
likelihood.train()

optimizer = torch.optim.Adam([
    {'params': list(model.parameters())[2:]},
], lr=learning_rate)

mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

for i in range(100):
    # Zero gradients from previous iteration
    optimizer.zero_grad()
    # Output from model
    output = model(X)
    # Calc loss and backprop gradients
    loss = -mll(output, y)
    loss.backward()
    if not i % 30:
        print('i: {0}, Loss: {1}'.format(i, loss.item()))
    optimizer.step()
```
Now, let's check predictions:

```
model.eval()
likelihood.eval()

with torch.no_grad(), gpytorch.settings.fast_pred_var():
    y_pred = likelihood(model(X)).mean.detach().numpy()

fig, ax = plt.subplots(figsize=(20,10))
ax.plot(data['x'], data['y'], 'o', color='red',label='training', markersize=5, fillstyle='none')
ax.plot(data['x'], y_pred, color='blue')
clean_ax(ax)
```

![image](https://user-images.githubusercontent.com/19956442/54079709-8119f380-4296-11e9-8d61-4025e9616af0.png)

This isn't what I expected. There is a lot more structure that the GP isn't discovering. So I have a few question, but let's also check out the hypers:

```
list(model.named_hyperparameters())
```

![image](https://user-images.githubusercontent.com/19956442/54079750-3c428c80-4297-11e9-98cb-85cf812eacd1.png)

With that:

1. The model seems to think the true function is smooth and we just have very noisy observations. How is this level of noise reflected in our hypers? The noise in the likelihood is 0 (intentionally) and the white noise kernel variances is at 0.02^2. Under this circumstance, shouldn't the GP do a lot more wiggly interpolation?
2. Since our predictive mean is formed as:
![image](https://user-images.githubusercontent.com/19956442/54079864-de637400-4299-11e9-8bfc-8ecfeed4be73.png)

Then our predictions decompose in the same fashion as our kernel. That is k^T_{Xx^*} is constructed in the same way we built out kernel. So, is there any way we can see these component pieces of our predictions? I've tried constructed this myself by tracing the code without much luck. It would be very useful for diagnosing _why_ a GP produces wonky results.

3. Do we just think this is a local vs global optimum issue with the hyperparameters? In a circumstance like this, would you say it's best to rely on hand selection of hypers?

Any advice you can provide (inside or outside these questions) is appreciate!

Thanks!
DJ

**Comment by jacobrgardner** [comment]
Hey DJ,

I'll take a look at the data and answer more fully, but a couple of things that pop to mind in the short term:

1. The parameter that is 0 is actually the raw_noise, which is the parameter before it is transformed to be positive by the softplus (which is accessible by looking at just likelihood.noise). Therefore the noise of the actual model is something like 0.7 (softplus(0))), which is actually quite high given the range of data. To set the actual noise:
```python
likelihood.initialize(noise=1e-4)
```

2. To make predictions with components of the kernel, what I would do is make your GP class definition itself do this in forward, and have forward take a second argument to control the kernel behavior instead of just calling the covar_module. This use case is one of the main reason the interface involves writing a custom subclass of ExactGP instead of using a prebaked one.

3. I doubt this is a local optimum issue. Most likely what's going on is my answer to #1.

**Comment by Duane321** [comment]
Thanks for the help here - appreciate it.

But I think I'm screwing up your advice on 2. I've rewritten my GP class as:

```
class GP(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(GP, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        
        spline_indices = [0, 1, 2, 3, 4, 5]
        self.ker_linear_t = gpytorch.kernels.LinearKernel(num_dimensions=len(spline_indices), 
                                                          active_dims=spline_indices)
        self.ker_linear_m = gpytorch.kernels.LinearKernel(num_dimensions=len(spline_indices), 
                                                          active_dims=spline_indices)
        
        self.ker_cyc_half = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(active_dims=[6,7]))
        self.ker_cyc_10th = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(active_dims=[8,9]))
        
        noise_vars = np.array([0.02**2]*len(x))
        noise_vars = torch.from_numpy(noise_vars).to(torch.float64)
        self.ker_noise = gpytorch.kernels.WhiteNoiseKernel(noise_vars)
        
self.ker_noise)
    
    def forward(self, x, components=False):
        
        mean_x = self.mean_module(x)
        ker_linear_t_cov = self.ker_linear_t(x)
        ker_linear_m_cov = self.ker_linear_m(x)
        ker_cyc_half_cov = self.ker_cyc_half(x)
        ker_cyc_10th_cov = self.ker_cyc_10th(x)
        ker_noise_cov = self.ker_noise(x)
        
        if components:
            component_covs = {}
            component_covs['ker_linear_t_cov'] = gpytorch.distributions.MultivariateNormal(mean_x, ker_linear_t_cov)
            component_covs['ker_linear_m_cov'] = gpytorch.distributions.MultivariateNormal(mean_x, ker_linear_m_cov)
            component_covs['ker_cyc_half_cov'] = gpytorch.distributions.MultivariateNormal(mean_x, ker_cyc_half_cov)
            component_covs['ker_cyc_10th_cov'] = gpytorch.distributions.MultivariateNormal(mean_x, ker_cyc_10th_cov)
            component_covs['ker_noise_cov'] = gpytorch.distributions.MultivariateNormal(mean_x, ker_noise_cov)
            return component_covs
        else:
            covar_x = ker_linear_m_cov*(ker_linear_t_cov + ker_cyc_half_cov + ker_cyc_10th_cov + ker_noise_cov)
            return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
```

The changes are:

1. I've moved the combinations of kernels into the forward method.
2. I've added an option to the forward the changes the outputs to be the kernel components.

Though this doesn't work. Those kernel components return a mean of zero in:

```
model.eval()
likelihood.eval()

with torch.no_grad(), gpytorch.settings.fast_pred_var():
        
    pred_components = model.forward(X, components=True)
    y_pred = {}
    for key, val in pred_components.items():
        y_pred[key] = likelihood(val).mean.detach().numpy()
```

So whatever I'm doing has caused this routine to miss the portion where we condition on y. Do you have any advice on what I'm doing wrong? Thanks!

**Comment by Duane321** [comment]
Thanks for the help!

Regarding 2) I think I'

**Comment by Duane321** [comment]
Sorry - accidental closing

**Comment by jacobrgardner** [comment]
Ah, you are running in to two problems here.

1. Manually calling forward directly calls the forward method you've defined, which just returns the prior:
```python
pred_components = model.forward(X, components=True)
```
Like other PyTorch models, the intended mechanism is to use the `__call__` method of the module directly like so:
```python
pred_components = model(X, components=True)
```

2. `ExactGP.forward` should always return a MultivariateNormal. What I would recommend doing is similar to what you have, but instead of returning the dict returning a `MultivariateNormal(mean_x, requested_covar_component)`. Then in eval mode you'll get the predictive posterior restricted to that component. For example, viewing the function you are constructing as `f(x) = f_{1}(x) + f_{2}(x) + f_{3}(x) + f_{4}(x)`, if you were to return:

```python
return MultivariateNormal(mean_x, ker_linear_m_cov * ker_cyc_half_cov)
```
Then calling the model in eval() mode would get you the predictive distribution for `f_{2}(x)`.

**Comment by Duane321** [comment]
Hey,

So I've since updated GPyTorch a few times (with the alpha version currently). I'm noticing that this decomposition doesn't exactly hold anymore. However, it is very correlated. It's _almost_ like a linear transformation of my reconstruction:

![image](https://user-images.githubusercontent.com/19956442/56474707-31e8e600-6432-11e9-95cf-2c448315083c.png)

In the past, I've checked that this relation holds exactly. What's changed? Are there any adjustments you can recommend such that I can reconstruct exactly? This decomposition is super useful for debugging predictions.

Thanks!

**Comment by gpleiss** [comment]
@Duane321 can you post your model?

**Comment by Duane321** [comment]
@gpleiss Certainly

```
class GP(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(GP, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        
        spline_indices = [0, 1, 2, 3, 4, 5]
        self.ker_linear_t = gpytorch.kernels.LinearKernel(num_dimensions=len(spline_indices), 
                                                          active_dims=spline_indices)
        self.ker_linear_m = gpytorch.kernels.LinearKernel(num_dimensions=len(spline_indices), 
                                                          active_dims=spline_indices)
        
        self.ker_cyc_half = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(active_dims=[6,7]))
        self.ker_cyc_10th = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(active_dims=[8,9]))
        
        noise_vars = np.array([0.02**2]*len(x))
        noise_vars = torch.from_numpy(noise_vars).to(torch.float64)
            
    def forward(self, x, requested_comp=None):
        
        mean_x = self.mean_module(x)
        ker_linear_t_cov = self.ker_linear_t(x)
        ker_linear_m_cov = self.ker_linear_m(x)
        ker_cyc_half_cov = self.ker_cyc_half(x)
        ker_cyc_10th_cov = self.ker_cyc_10th(x)
        
        if requested_comp is None:
            covar_x = ker_linear_m_cov*(ker_linear_t_cov + ker_cyc_half_cov + ker_cyc_10th_cov)
        elif requested_comp == 'ker_linear_t_cov':
            covar_x = ker_linear_t_cov
        elif requested_comp == 'ker_linear_m_cov':
            covar_x = ker_linear_m_cov
        elif requested_comp == 'ker_cyc_half_cov':
            covar_x = ker_cyc_half_cov
        elif requested_comp == 'ker_cyc_10th_cov':
            covar_x = ker_cyc_10th_cov
        else:
            raise ValueError('Given bad requested_comp.')
        
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
```

Then, I produce my attempted reconstruction this way:

```
model.eval()
likelihood.eval()
pred_components = {}
with torch.no_grad():
        
    y_pred = likelihood(model(X)).mean.detach().numpy()
        
    for ker in ['ker_linear_t_cov', 'ker_linear_m_cov', 
                'ker_cyc_half_cov', 'ker_cyc_10th_cov']:
        pred_components[ker] = likelihood(model(X,requested_comp=ker)).mean.numpy()

y_pred_2 = pred_components['ker_linear_m_cov']*(pred_components['ker_linear_t_cov'] + pred_components['ker_cyc_half_cov'] + pred_components['ker_cyc_10th_cov'])

compare = pd.DataFrame({'y_pred':y_pred,'my_reconstruction':y_pred_2})
```

And then the issue is, the columns in `compare` aren't the same. This is true even when the mean_constant parameter is set to zero.

**Comment by jacobrgardner** [comment]
@Duane321 Do you have a commit hash where this *did* work? This would allow us to use `git bisect` to see exactly what changes happened and what you'll need to adjust to get back this behavior.

**Comment by Duane321** [comment]
Hey @jacobrgardner, so I gave this a shot, but couldn't find the hash. Whenever I checked out an older version of the master branch, I ran into this error:

![image](https://user-images.githubusercontent.com/19956442/56712770-7ac7c580-66e4-11e9-8a91-61fae5b95997.png)

Which doesn't make much sense to me - I must be failing to recreate my own history somehow.

Regardless, do we think the above method shouldn't work? It's very close, but maybe this is no longer the best way?

**Comment by jacobrgardner** [comment]
Possibly. Without the commit hash I'll need to just actually go through your code and see what you are doing, which will just take a bit longer -- I'll try to get around to it as soon as I can

**Comment by Duane321** [comment]
No worries @jacobrgardner you've helped me a ton already. I'll give finding the commit hash another go within the next few days. I should be able to find it.

**Comment by gpleiss** [comment]
closing for now - reopen if there's still questions :)
